File: src/domain/pilot/value-objects/sync-status.ts
// src/domain/pilot/value-objects/sync-status.ts

import * as S from "effect/Schema"
import { ShopifyProductIdSchema, type ShopifyProductId } from "./ids"

// ============================================
// SYNC ERROR
// ============================================

export const SyncErrorSchema = S.Struct({
  code: S.String,
  message: S.String,
  details: S.Unknown,
})

export type SyncError = typeof SyncErrorSchema.Type

// ============================================
// NOT SYNCED
// ============================================

const NotSyncedSchema = S.TaggedStruct("NotSynced", {})

export type NotSynced = typeof NotSyncedSchema.Type

// ============================================
// SYNCED
// ============================================

const SyncedSchema = S.TaggedStruct("Synced", {
  shopifyProductId: ShopifyProductIdSchema,
  syncedAt: S.Date,
})

export type Synced = typeof SyncedSchema.Type

// ============================================
// SYNC FAILED
// ============================================

const SyncFailedSchema = S.TaggedStruct("SyncFailed", {
  error: SyncErrorSchema,
  failedAt: S.Date,
  attempts: S.Number,
})

export type SyncFailed = typeof SyncFailedSchema.Type

// ============================================
// SYNC STATUS (union)
// ============================================

export const SyncStatusSchema = S.Union(NotSyncedSchema, SyncedSchema, SyncFailedSchema)

export type SyncStatus = typeof SyncStatusSchema.Type

// ============================================
// FACTORIES
// ============================================

export const MakeNotSynced = (): NotSynced => ({ _tag: "NotSynced" })

export const MakeSynced = (params: {
  shopifyProductId: ShopifyProductId
  syncedAt: Date
}): Synced => ({
  _tag: "Synced",
  ...params,
})

export const MakeSyncFailed = (params: {
  error: SyncError
  failedAt: Date
  attempts: number
}): SyncFailed => ({
  _tag: "SyncFailed",
  ...params,
})

File: apps/server-java/src/main/java/com/maisonamane/domain/pilot/valueobject/SyncStatus.java
package com.maisonamane.domain.pilot.valueobject;

import java.time.Instant;
import java.util.Objects;

/**
 * Sealed interface representing the sync status state machine.
 * Maps Effect-TS S.Union(NotSynced, Synced, SyncFailed) to Java sealed interface.
 */
public sealed interface SyncStatus permits SyncStatus.NotSynced, SyncStatus.Synced, SyncStatus.SyncFailed {

    record NotSynced() implements SyncStatus {
        private static final NotSynced INSTANCE = new NotSynced();

        public static NotSynced instance() {
            return INSTANCE;
        }
    }

    record Synced(ShopifyProductId shopifyProductId, Instant syncedAt) implements SyncStatus {
        public Synced {
            Objects.requireNonNull(shopifyProductId, "ShopifyProductId cannot be null");
            Objects.requireNonNull(syncedAt, "SyncedAt cannot be null");
        }

        public static Synced of(ShopifyProductId shopifyProductId, Instant syncedAt) {
            return new Synced(shopifyProductId, syncedAt);
        }
    }

    record SyncFailed(SyncError error, Instant failedAt, int attempts) implements SyncStatus {
        public SyncFailed {
            Objects.requireNonNull(error, "Error cannot be null");
            Objects.requireNonNull(failedAt, "FailedAt cannot be null");
            if (attempts < 1) {
                throw new IllegalArgumentException("Attempts must be at least 1");
            }
        }

        public static SyncFailed of(SyncError error, Instant failedAt, int attempts) {
            return new SyncFailed(error, failedAt, attempts);
        }
    }

    // Factory methods
    static NotSynced notSynced() {
        return NotSynced.instance();
    }

    static Synced synced(ShopifyProductId shopifyProductId, Instant syncedAt) {
        return Synced.of(shopifyProductId, syncedAt);
    }

    static SyncFailed syncFailed(SyncError error, Instant failedAt, int attempts) {
        return SyncFailed.of(error, failedAt, attempts);
    }

    // Type checks
    default boolean isNotSynced() {
        return this instanceof NotSynced;
    }

    default boolean isSynced() {
        return this instanceof Synced;
    }

    default boolean isSyncFailed() {
        return this instanceof SyncFailed;
    }
}
